src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -310,50 +311,27 @@ private async Task DequeueBlocksContinuouslyAsync()
                     this.logger.LogDebug("Node is shutting down. Save batch.");
                 }
 
-                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip
-                // or the max batch size is reached or the node is shutting down.
-                if (dequeueTask.Status == TaskStatus.RanToCompletion)
-                {
-                    ChainedHeaderBlock item = dequeueTask.Result;
-
-                    // Set the dequeue task to null so it can be assigned on the next iteration.
-                    dequeueTask = null;
-
-                    lock (this.getBlockLock)
-                    {
-                        this.batch.Add(item);
-                    }
-
-                    this.currentBatchSizeBytes += item.Block.BlockSize.Value;
-
-                    saveBatch = saveBatch || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes) || this.chainState.IsAtBestChainTip;
-                }
-                else
-                {
-                    // Will be executed in case timer ran out or node is being shut down.
-                    saveBatch = true;
-                }
+                saveBatch = saveBatch || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes) || this.chainState.IsAtBestChainTip;
 
                 if (saveBatch)
                 {
                     if (this.batch.Count != 0)
                     {
-                        await this.SaveBatchAsync().ConfigureAwait(false);
+                        List<ChainedHeaderBlock> saved = await this.SaveBatchAsync().ConfigureAwait(false);
 
                         lock (this.getBlockLock)
                         {
-                            this.batch.Clear();
-                        }
+                            // Only remove items from the batch that have explicitly been saved.
+                            foreach (ChainedHeaderBlock savedHeaderBlock in saved)
```

Comment:
this will lead to race condition when we would remove stuff from a batch that we need. 
in case we are connecting a chain and then disconnecting it because it's invalid and reconnecting old one it's possible that we will have 2 instances of same block in the batch but it will be removed here. 
That will lead to the absence in the block which is a part of the best chain

Reply:
If so, it seems to me it was a pre-existing race condition. The previous implementation would clear the entire batch (which at that point would contain reorged blocks) after saving.

I can try rectify your concern as part of this PR if you want, or I can do it as part of implementing additional integration tests. This seems like a good test candidate.